File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/content/form/inputtype/geo/GeoPoint.ts
module api.content.form.inputtype.geo {

    import support = api.form.inputtype.support;

    import ValueTypes = api.data.type.ValueTypes;

    export class GeoPoint extends support.BaseInputTypeNotManagingAdd<any,api.util.GeoPoint> {

        constructor(config: api.form.inputtype.InputTypeViewContext<any>) {
            super(config);
        }

        getValueType(): api.data.type.ValueType {
            return ValueTypes.GEO_POINT;
        }

        newInitialValue(): api.util.GeoPoint {
            return null;
        }

        createInputOccurrenceElement(index: number, property: api.data.Property): api.dom.Element {

            var geoPoint = new api.ui.geo.GeoPoint();

            if (property.hasNonNullValue()) {
                var geoPointValue = property.getGeoPoint();
                if (geoPointValue) {
                    geoPoint.setGeoPoint(geoPointValue);
                }
            }
            return geoPoint;
        }

        availableSizeChanged() {
        }

        onOccurrenceValueChanged(element: api.dom.Element, listener: (event: api.form.inputtype.support.ValueChangedEvent) => void) {

            var geoPoint = <api.ui.geo.GeoPoint>element;

            geoPoint.onValueChanged((event: api.ui.ValueChangedEvent) => {
                var geoLocation = event.getNewValue();
                var value = ValueTypes.GEO_POINT.newValue(geoLocation);
                listener(new api.form.inputtype.support.ValueChangedEvent(value));
            });


        }

        getValue(occurrence: api.dom.Element): api.data.Value {
            var geoPointEL: api.ui.geo.GeoPoint = <api.ui.geo.GeoPoint>occurrence;
            var geoPointVo: api.util.GeoPoint = geoPointEL.getGeoPoint();
            return  new api.data.Value(geoPointVo, ValueTypes.GEO_POINT);
        }

        valueBreaksRequiredContract(value: api.data.Value): boolean {
            return value.isNull() || !value.getType().equals(ValueTypes.GEO_POINT);
        }
    }

    api.form.inputtype.InputTypeManager.register(new api.Class("GeoPoint", GeoPoint));
}

File: modules/core-api/src/main/java/com/enonic/xp/core/form/inputtype/GeoPoint.java
package com.enonic.xp.core.form.inputtype;

import com.enonic.xp.core.data.Property;
import com.enonic.xp.core.data.Value;
import com.enonic.xp.core.form.BreaksRequiredContractException;

final class GeoPoint
    extends InputType
{
    GeoPoint()
    {
    }

    @Override
    public void checkBreaksRequiredContract( final Property property )
        throws BreaksRequiredContractException
    {
        final com.enonic.xp.core.util.GeoPoint value = property.getGeoPoint();
        if ( value == null )
        {
            throw new BreaksRequiredContractException( property, this );
        }
    }

    @Override
    public Value newValue( final String value )
    {
        return Value.newGeoPoint( com.enonic.xp.core.util.GeoPoint.from( value ) );
    }

    @Override
    public InputTypeConfig getDefaultConfig()
    {
        return null;
    }

}
